pkg/policy/mapstate.go

```diff
@@ -142,7 +157,76 @@ func (e MapStateEntry) String() string {
 
 // DenyPreferredInsert inserts a key and entry into the map by given preference
 // to deny entries, and L3-only deny entries over L3-L4 allows.
+// This form may be used when a full policy is computed and we are not yet interested
+// in accumulating incremental changes.
 func (keys MapState) DenyPreferredInsert(newKey Key, newEntry MapStateEntry) {
+	keys.denyPreferredInsertWithChanges(newKey, newEntry, nil, nil)
+}
+
+// addKeyWithChanges adds a 'key' with value 'entry' to 'keys' keeping track of incremental changes in 'adds' and 'deletes'
+func (keys MapState) addKeyWithChanges(key Key, entry MapStateEntry, adds, deletes MapState) {
+	// Keep all selectors that need this entry so that it is deleted only if all the selectors delete their contribution
+	updatedEntry := entry
+	oldEntry, exists := keys[key]
+	if exists {
+		// keep the existing selectors map of the old entry
+		updatedEntry.selectors = oldEntry.selectors
+	} else if len(entry.selectors) > 0 {
+		// create a new selectors map
+		updatedEntry.selectors = make(map[CachedSelector]struct{}, len(entry.selectors))
+	} else {
+		// Keep the map as nil when empty. This makes a difference for unit testing only.
+		updatedEntry.selectors = nil
+	}
+
+	// TODO: Do we need to merge labels as well?
+	// Merge new selectors to the updated entry without modifying 'entry' as it is being reused by the caller
+	updatedEntry.MergeSelectors(&entry)
+	// Update (or insert) the entry
+	keys[key] = updatedEntry
+
+	// Record an incremental Add if desired and entry is new or changed
+	if adds != nil && (!exists || entry.IsDeny != oldEntry.IsDeny || entry.ProxyPort != oldEntry.ProxyPort) {
```

Comment:
Do we have a convention to name a function that is testing equivalence rather than equality? In this case we are interested in equivalence from bpf map viewpoint, ignoring all user-space-only bookkeeping.

Reply:
I later realized that you check individual fields in other forms so this is actually really minor because there are more open coded checks anyway.

Reply:
Figured that this exists:
```
pkg/policy/mapstate.go:// DatapathEqual returns true of two entries are equal in the datapath's PoV,
pkg/policy/mapstate.go:func (e *MapStateEntry) DatapathEqual(o *MapStateEntry) bool {
```